src/iter/mod.rs

```diff
@@ -2417,6 +2421,82 @@ pub trait IndexedParallelIterator: ParallelIterator {
         Chunks::new(self, chunk_size)
     }
 
+    /// Splits an iterator into fixed-size chunks, performing a sequential [`fold()`] on
+    /// each chunk.
+    ///
+    /// Returns an iterator that produces a folded result for each chunk of items
+    /// produced by this iterator.
+    ///
+    /// This works essentially like:
+    ///
+    /// ```text
+    /// iter.chunks(chunk_size)
+    ///     .map(|chunk|
+    ///         chunk.into_iter()
+    ///             .fold(identity, fold_op)
+    ///     )
+    /// ```
+    ///
+    /// except there is no per-chunk allocation overhead.
+    ///
+    /// [`fold()`]: std::iter::Iterator#method.fold
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use rayon::prelude::*;
+    /// let nums = vec![1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
+    /// let chunk_sums = nums.into_par_iter().fold_chunks(2, || 0, |a, n| a + n).collect::<Vec<_>>();
+    /// assert_eq!(chunk_sums, vec![3, 7, 11, 15, 19]);
+    /// ```
+    fn fold_chunks<ID, F, U>(
+        self,
+        chunk_size: usize,
+        identity: ID,
+        fold_op: F,
+    ) -> FoldChunks<Self, ID, F>
+    where
+        ID: Fn() -> U + Send + Sync,
+        F: Fn(U, Self::Item) -> U + Send + Sync,
+    {
+        assert!(chunk_size != 0, "chunk_size must not be zero");
```

Comment:
I think it would be good to mention the panic here and below in the doc comment.

Reply:
And maybe `#[track_caller]` to follow #977.

Reply:
Thanks for taking a look!

> Note that the `join` and `par_iter` versions are converting every number to `BigUint`, causing extra memory allocation overhead, while your fold is multiplying `BigUint * u32`....
> ...
> However, adding `.with_min_len(64)` to mine does get its result consistently within noise of yours.

Explains my surprise (should have noticed that!)

I'll push my factorial benchmark here shortly. To address the 'like-for-like' issue should it match the `join` and `par_iter` version by mapping everything to `BigUint` first? E.g.
```rust
            .into_par_iter()
            .map(BigUint::from)
            .fold_chunks_with(64, BigUint::from(1_u32), |acc, x| acc.mul(x))
            ...
```
... or maybe supplement with the fastest version your `factorial_fold_with` instead?

Reply:
> I suspect this kind of chunking is probably easier to understand than the loose `with_min_len`/`with_max_len` controls, even though those are more flexible.

Yes, and also with `fold_chunks` or `chunks` you can guarantee something like: 'all chunks have size _x_ except for the last chunk which may be smaller', a common requirement for protocols/standards. Whereas, if I understand correctly, using `with_min_len`/`with_max_len` controls would not provide such a guarantee(?).

Reply:
Also, it would be nice to compare against a version using `par_iter().chunks()...` since this was the original motivation for `fold_chunks`: e.g.
```rust
#[bench]
/// Compute the Factorial using rayon::chunks/map and a sequential fold.
fn factorial_chunks_fold(b: &mut test::Bencher) {
    println!("factorial_fold_chunks, chunk size: {}", N);
    fn fact(n: u32) -> BigUint {
        (1..n + 1)
            .into_par_iter()
            .chunks(64)
            .map(|chunk| chunk.into_iter().fold(BigUint::from(1_u32), |acc, x| acc.mul(x)))
            .reduce_with(Mul::mul)
            .unwrap()
    }

    let f = factorial(N);
    b.iter(|| assert_eq!(fact(test::black_box(N)), f));
}
```
However the performance is within noise of `fold_chunks` and the difference seen in the criterion benchmark results above is not seen (use of `BigUint` masks it?). I think it would need to be a whole new benchmark module which didn't require allocation in its 'core' operation, something like summing an iterator of `u32`s into a `u64` perhaps.

Reply:
Agreed, however there's quite a few places that do the same without mentioning in the doc comment (`IndexedParallelIterator::chunks` and various methods on `ParallelSlice`). Perhaps it would be a cleaner git history to address all of these together in their own PR?